crates/astria-sequencer/src/app.rs

```diff
@@ -0,0 +1,195 @@
+use std::sync::Arc;
+
+use anyhow::Result;
+use penumbra_component::Component;
+use penumbra_storage::{
+    ArcStateDeltaExt,
+    Snapshot,
+    StateDelta,
+    Storage,
+};
+use tendermint::abci::{
+    self,
+    Event,
+};
+use tracing::info;
+
+use crate::{
+    accounts::component::AccountsComponent,
+    state_ext::{
+        StateReadExt as _,
+        StateWriteExt as _,
+    },
+    transaction::Transaction,
+};
+
+/// The application hash, used to verify the application state.
+/// TODO: this may not be the same as the state root hash?
+pub type AppHash = penumbra_storage::RootHash;
+
+/// The inter-block state being written to by the application.
+type InterBlockState = Arc<StateDelta<Snapshot>>;
```

Comment:
this was mirrored from penumbra: https://github.com/penumbra-zone/penumbra/blob/9cc2c644e05c61d21fdc7b507b96016ba6b9a935/app/src/app/mod.rs#L35

Reply:
I believe it's wrapped in an `Arc` because multiple components need to take ownership of it within one function to mutate it (ie write to the `StateDelta`)

Reply:
it is also relatively short-lived; it only exists within this current block, at the end of the block it gets replaced (see `App.commit`)

```
        // Get the latest version of the state, now that we've committed it.
        self.state = Arc::new(StateDelta::new(storage.latest_snapshot()));
```

Reply:
Gotcha. Thank you for pointing to that. This design surprised me because usage `StateDelta` through `ArcStateDeltaExt` requires exclusive access. Since there are no other protections (like Mutex) it seems like a big burden on the developer to ensure that weak/strong references are always dropped in time.